### ISSUE TITLE
Fix for deferred_init in Task.init is incorrect

### ISSUE THREAD

**Issue by idantene** [allegroai/clearml#855]
## Describe the bug
I'm not 100% sure, but we noticed this is an aside in our flow.
In [this commit](https://github.com/allegroai/clearml/commit/1729b9019f047eaadb16ee8a77cbe99ddb3290ef), the applied change does not produce the desired results AFAICS.

Original had `deferred_init != 0` (where `deferred_init` is a boolean argument for `Task.init`), and the updated has `not (type(deferred_init) == int and deferred_init == 0)`.
This can also be written as `type(deferred_init) != int or deferred_init != 0`, which then always evaluates to `True` for the earlier clause (since the input argument is boolean).

One must wonder, in this specific case, why not compare the boolean to booleans with e.g. `deferred_init is True`?

## To reproduce
No real useful code to reproduce, but simply:
```python3
deferred_init = False
deferred_init != 0  # False
not (type(deferred_init) == int and deferred_init == 0)  # True
```

## Environment
* ClearML SDK Version
1.8.3
* Python Version
3.10

**Comment by erezalg** [comment]
Hi @idantene,

Thanks for raising this! We looked at the code and while it's working correctly (At least as far as we could tell :) ), it is indeed confusing, we plan to simplify it in the next release

**Comment by idantene** [comment]
Hey @erezalg,

I'm not sure how it's working as intended in that part of the code.
I understand elsewhere you take the value from an environment variable which is either 0 or 1, but here the value is expected to be a boolean given in the arguments to `Task.init`.
As it is right now, whether you pass `deferred_init=True` or `deferred_init=False` does not matter, since the type test fails.

**Comment by idantene** [comment]
Maybe as an additional clarification @erezalg, we use the default argument here (that is, we do **not** defer the initialization of tasks).
However, because of this flawed condition, the `Task.init(...)` call goes into that clause and crashes our flow. This did not happen earlier (the condition was also flawed, but `deferred_init != 0` at least evaluates correctly).

**Comment by erezalg** [comment]
Hi @idantene,

Ok, I think I understand now and I think we fixed this and it should be fixed in v1.9
We'll update here once something is ready :)

**Comment by idantene** [comment]
We'd be happy to help test a release candidate once it's available 👌

**Comment by idantene** [comment]
Last but not least - the error was introduced when renaming `wait_for_task_init` to `deferred_init` [here](https://github.com/allegroai/clearml/commit/556e9b25fe2166134db7f82db19425ddfac474b2).
Notice that originally (`wait_for_task_init`), all tests rely on boolean validation and pass a boolean value, whereas in the transition, the confusion between booleans and integers was added (default argument remained a boolean, but passed arguments within the call use an integer value).

**Comment by idantene** [comment]
@erezalg I don't think the fix fixes this either... 😓

**Comment by erezalg** [comment]
Hi @idantene,

Weird, I think it works for me with 1.9.1rc0 (Which AFAIK, has nothing in it that dealt with deffered init)
I ran:
from clearml import Task
import time

print('before init')
start_time = time.time()
t = Task.init(project_name='erez-test', task_name='test deferred', deferred_init=True, reuse_last_task_id=False)
print("Took", time.time() - start_time)

And got:
before init
ClearML initializing Task in the background
Took 0.00078582763671875

Which to my understanding means deferred init works, let me know if I got it incorrectly!

**Comment by idantene** [comment]
That is true :)
I guess somehow this breaks our current functionality somehow though - I'll have to figure that one out myself I suppose.

Thanks!

**Comment by idantene** [comment]
Actually, maybe it's a good idea to open this up anyway.

We have not yet transitioned to pipelines (we started using ClearML before it was available, and we have some technical debts...). We noticed that one cannot have a remotely-running task start new tasks, but one could've done so locally. So we've overridden the variable that sets a task as "remote", with
```python3
clearml.config.remote.override_current_task_id(None)
```
And now our remote task can execute more tasks freely, hurray.

Now to the issue at hand, the following code works nicely with `clearml==1.8.0`, but crashes with newer versions:
```python3
import clearml
from clearml import Task
import time

print('before init')
start_time = time.time()
t = Task.init(project_name='internal tests', task_name='test deferred', deferred_init=False, reuse_last_task_id=False)
t.execute_remotely(queue_name='default')
print("Took", time.time() - start_time)
clearml.config.remote.override_current_task_id(None)
t0 = time.time()
#t.close()
t2 = Task.init(project_name='internal tests', task_name='sub task', deferred_init=False, reuse_last_task_id=False)
print("Took", time.time() - t0)
```

With the newer versions, the error message is:
```
clearml.errors.UsageError: Current task already created and requested task name 'sub task' does not match current task name 'test deferred'. If you wish to create additional tasks use `Task.create`, or close the current task with `task.close()` before calling `Task.init(...)`
```

And that's basically how I ended up with this issue :) - since the only thing that changed was the use of `defered_init` (and how that condition is used).

I'm not sure what else changed, maybe how the `Task.__main_task` is set?

**Comment by idantene** [comment]
Would it be safe to call e.g. `t._reset_current_task_obj()`?

**Comment by erezalg** [comment]
@idantene,

Thanks for the detailed write-up, let me see if I get you correctly :)

You're trying to have a remote task start more tasks and you can't so you "hacked" it. Can you elaborate what you tried and what didn't work? Although it's not related to this issue, maybe we can (and should) fix this but I'd like to understand what you tried first.

I tried your code and indeed something broke, not really sure why, and not sure it's related to deferred init (as it's set to false anyway). I will provide this to the dev team and see what did we change

**Comment by idantene** [comment]
> @idantene,
> 
> Thanks for the detailed write-up, let me see if I get you correctly :)
> 
> You're trying to have a remote task start more tasks and you can't so you "hacked" it. Can you elaborate what you tried and what didn't work? Although it's not related to this issue, maybe we can (and should) fix this but I'd like to understand what you tried first.
> 
> I tried your code and indeed something broke, not really sure why, and not sure it's related to deferred init (as it's set to false anyway). I will provide this to the dev team and see what did we change

Not sure what else I can provide :) This "hack" worked for us perfectly from ClearML v1.00 up until v1.8.1 where it breaks.
One simply needs to ensure that ClearML does not thing this remote task is indeed remote, and then it can spawn more tasks as needed - hence the call to `clearml.config.remote.override_current_task_id(None)`.
We now have to specifically close this spawning task, which hinders our original goal of monitoring all the tasks as needed in a single "parent" task.

Did you try the above code with ClearML v1.8.0?

**Comment by erezalg** [comment]
@idantene,

Sorry I think I wasn't clear. Yes I reproduced your bug and I gave it to the dev team to figure out what is happening. I'll update here once we figure it out.

I just wanted to understand, why did you resort to this hack:
clearml.config.remote.override_current_task_id(None)

**Comment by idantene** [comment]
Ah! That was an interesting road down ClearML code.
I landed here - https://github.com/allegroai/clearml/blob/master/clearml/task.py#L559, and immediately noticed the else condition, whereby any `Task.init` in a remote-execution is almost a no-op (no initialization) - https://github.com/allegroai/clearml/blob/master/clearml/task.py#L635.

That meant I had to see how to make sure `running_remotely` is `False`, which led me [here](https://github.com/allegroai/clearml/blob/master/clearml/config/__init__.py#L130) and then [here](https://github.com/allegroai/clearml/blob/master/clearml/config/remote.py#L6).

**Comment by erezalg** [comment]
@idantene,

One other question that popped up, because, TBH, we didn't expect calling Task.init() twice to work :smile: why aren't you using Task.create() for the spawned tasks?

**Comment by idantene** [comment]
I can't recall the specifics, but it was vaguely to do with the auto-magic components of it all.
